File: js/src/status.ts
/**
 * User-facing one-line status banners, printed to stderr so piping stdout
 * stays clean. Purely additive: never touches wire behaviour, reconnect
 * timing, or existing logger output.
 *
 * Wording here MUST match `sdk/python/aurival`'s equivalent module
 * character-for-character (SDK-19) — only the language differs.
 *
 * Off by default with `Bot({ quiet: true })` or `AURIVAL_QUIET=1` (either
 * silences every line in this module).
 */

const ANSI = {
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  red: '\x1b[31m',
  reset: '\x1b[0m',
} as const;

type Color = keyof typeof ANSI;

/** Env var wins alongside the constructor flag — either silences everything. */
export function isQuiet(optionQuiet: boolean | undefined): boolean {
  if (optionQuiet) return true;
  return process.env['AURIVAL_QUIET'] === '1';
}

function colorize(line: string, color: Color | null): string {
  if (color === null) return line;
  if (!process.stderr.isTTY) return line;
  return `${ANSI[color]}${line}${ANSI.reset}`;
}

function emit(line: string, color: Color | null, quiet: boolean): void {
  if (quiet) return;
  console.error(colorize(line, color));
}

/**
 * Human-friendly elapsed time, no external deps: `"2.3s"`, `"1m 5s"`,
 * `"1h 5m"`. Used for both `reconnected after <duration>` and
 * `disconnected after <uptime>`.
 */
export function formatDuration(ms: number): string {
  const totalSeconds = Math.max(0, ms) / 1000;
  if (totalSeconds < 60) {
    const rounded = Math.round(totalSeconds * 10) / 10;
    return `${rounded}s`;
  }
  const totalMinutes = Math.floor(totalSeconds / 60);
  const seconds = Math.round(totalSeconds - totalMinutes * 60);
  if (totalMinutes < 60) {
    return `${totalMinutes}m ${seconds}s`;
  }
  const hours = Math.floor(totalMinutes / 60);
  const minutes = totalMinutes % 60;
  return `${hours}h ${minutes}m`;
}

/** Same rounding as `formatDuration`'s sub-minute branch, without the unit. */
function formatSeconds(seconds: number): string {
  return `${Math.round(Math.max(0, seconds) * 10) / 10}`;
}

// -- one function per moment (SDK status banner spec) -----------------------

/** Once, at startup, before the first ws connect. */
export function connecting(handleOrId: string, quiet: boolean): void {
  emit(`aurival: connecting to bots.aurival.com as ${handleOrId}…`, 'yellow', quiet);
}

/** On receiving the very first `hello` frame for this run. */
export function connected(
  botName: string,
  sessionId: string,
  commandCount: number,
  quiet: boolean,
): void {
  const shortId = sessionId.slice(0, 8);
  emit(
    `aurival: connected — ${botName}, session ${shortId}, ${commandCount} commands registered. ` +
      `Waiting for commands. (Ctrl+C to stop)`,
    'green',
    quiet,
  );
}

/** On any close leading to a backoff/short-wait sleep before retry. */
export function reconnecting(reasonOrCode: string, delaySeconds: number, quiet: boolean): void {
  emit(
    `aurival: connection closed (${reasonOrCode}), reconnecting in ${formatSeconds(delaySeconds)}s…`,
    'yellow',
    quiet,
  );
}

/** On a `hello` that is NOT the very first one for this bot's run. */
export function reconnected(durationMs: number, quiet: boolean): void {
  emit(`aurival: reconnected after ${formatDuration(durationMs)}`, 'green', quiet);
}

/** Friendlier one-line text for the two fatal codes a bot operator hits by
 * ordinary use (a second connection, a pause from the app) rather than by
 * misconfiguration. Every other code keeps the generic `code: message` form. */
const FRIENDLY_STOPPED: Readonly<Record<string, string>> = {
  session_superseded:
    'another copy of this bot connected (elsewhere). One socket per bot: stop the other copy, then start this one again.',
  bot_suspended: 'this bot is paused by its owner. Resume it from the app, then start again.',
};

/** When the bye-action table resolves to `raise`, right before it propagates. */
export function stopped(code: string, message: string, quiet: boolean, docUrl?: string): void {
  const friendly = FRIENDLY_STOPPED[code];
  if (friendly !== undefined) {
    let line = `aurival: stopped — ${friendly}`;
    if (docUrl !== undefined) line += `\n${docUrl}`;
    emit(line, 'red', quiet);
    return;
  }
  emit(`aurival: stopped — ${code}: ${message}`, 'red', quiet);
}

/** A second `command(...)` registration for the same name at startup — the
 * overwrite silently wins today, and this makes it visible before it costs
 * someone a debugging session (SDK ergonomics, not a wire concern). */
export function duplicateCommand(name: string, quiet: boolean): void {
  emit(`aurival: command "${name}" registered twice, the later definition wins`, 'yellow', quiet);
}

/** Before the first connect, when zero commands were ever registered — a bot
 * that will sit there forever with nothing to do, printed while there is
 * still time to notice and add one. */
export function noCommands(quiet: boolean): void {
  emit(
    'aurival: no commands registered, this bot will connect and wait forever. Add @bot.command(...) before run().',
    'yellow',
    quiet,
  );
}

/** SIGINT/SIGTERM leading to a clean shutdown. */
export function disconnected(uptimeMs: number, quiet: boolean): void {
  emit(`aurival: disconnected after ${formatDuration(uptimeMs)}`, null, quiet);
}

File: python/aurival/status.py
"""Human-friendly one-line status banners, printed to stderr by default.

Six moments (connecting / connected / reconnecting / reconnected / stopped /
disconnected) so a developer who ran `python bot.py` sees SOMETHING rather
than a blank terminal. Always stderr — never stdout, so piping stays clean.
Colored only when stderr is a real TTY; plain text otherwise (piped,
redirected, or explicitly disabled).

Opt-out: `Bot(quiet=True)` or env `AURIVAL_QUIET=1` — either silences every
line this module prints. Checked live on every call (not cached at
construction) so an env var set after the fact still takes effect.
"""

from __future__ import annotations

import os
import sys
from typing import TextIO

_GREEN = "\033[32m"
_YELLOW = "\033[33m"
_RED = "\033[31m"
_RESET = "\033[0m"


def env_quiet() -> bool:
    return os.environ.get("AURIVAL_QUIET") == "1"


def format_duration(seconds: float) -> str:
    """Plain human-friendly duration, no external deps: `2.3s`, `1m 5s`, `1h 2m`."""
    seconds = max(0.0, seconds)
    if seconds < 60:
        return f"{seconds:.1f}s"
    total = int(round(seconds))
    minutes, sec = divmod(total, 60)
    if minutes < 60:
        return f"{minutes}m {sec}s"
    hours, minutes = divmod(minutes, 60)
    return f"{hours}h {minutes}m"


def _use_color(stream: TextIO) -> bool:
    try:
        return bool(stream.isatty())
    except Exception:
        return False


# Codes whose bare `{code}: {message}` reads like a crash rather than a
# designed stop — a second connection, or an owner pausing the bot from the
# app. Friendlier text, plus the doc_url on its own line so the developer can
# read the full story. Every other code is untouched.
_FRIENDLY_STOPPED: dict[str, str] = {
    "session_superseded": (
        "another copy of this bot connected (elsewhere). One socket per "
        "bot: stop the other copy, then start this one again."
    ),
    "bot_suspended": (
        "this bot is paused by its owner. Resume it from the app, then start again."
    ),
}


class StatusReporter:
    """One reporter per `Bot`, held by the bot and shared into its `Socket`
    so both sides of the connecting/connected/reconnecting story print
    through the same quiet/color rules."""

    def __init__(self, *, quiet: bool = False, stream: TextIO | None = None) -> None:
        self._quiet = quiet
        self._stream = stream if stream is not None else sys.stderr

    @property
    def enabled(self) -> bool:
        return not self._quiet and not env_quiet()

    def _emit(self, text: str, color: str | None) -> None:
        if not self.enabled:
            return
        stream = self._stream
        if color is not None and _use_color(stream):
            stream.write(f"{color}{text}{_RESET}\n")
        else:
            stream.write(f"{text}\n")
        stream.flush()

    def connecting(self, bot: str) -> None:
        self._emit(f"aurival: connecting to bots.aurival.com as {bot}…", _YELLOW)

    def connected(self, *, bot: str, session_id: str, command_count: int) -> None:
        short_id = session_id[:8]
        self._emit(
            f"aurival: connected — {bot}, session {short_id}, {command_count} commands "
            f"registered. Waiting for commands. (Ctrl+C to stop)",
            _GREEN,
        )

    def reconnecting(self, *, reason: str, delay_s: float) -> None:
        self._emit(f"aurival: connection closed ({reason}), reconnecting in {delay_s:.1f}s…", _YELLOW)

    def reconnected(self, *, duration_s: float) -> None:
        self._emit(f"aurival: reconnected after {format_duration(duration_s)}", _GREEN)

    def stopped(self, *, code: str, message: str, doc_url: str | None = None) -> None:
        friendly = _FRIENDLY_STOPPED.get(code)
        if friendly is not None:
            text = f"aurival: stopped — {friendly}"
            if doc_url is not None:
                text = f"{text}\n{doc_url}"
            self._emit(text, _RED)
            return
        self._emit(f"aurival: stopped — {code}: {message}", _RED)

    def duplicate_command(self, name: str) -> None:
        self._emit(
            f'aurival: command "{name}" registered twice, the later definition wins',
            _YELLOW,
        )

    def no_commands(self) -> None:
        self._emit(
            "aurival: no commands registered, this bot will connect and wait "
            "forever. Add @bot.command(...) before run().",
            _YELLOW,
        )

    def disconnected(self, *, duration_s: float) -> None:
        self._emit(f"aurival: disconnected after {format_duration(duration_s)}", None)
